### DISCUSSION TITLE
Numbering survey questions (6.3.1)

### DISCUSSION THREAD

**Discussion by fspinar** [jspsych/jsPsych#2794] [Q&A]
Hi all, 

I have a set of survey questions and I want to be able to enumerate them, and the numbering should be random. Besides, the survey will be presented twice during the experiment, with two separate random orderings.

So, I need a way to add question numbers to questions in a way that allows me to change question prompts. I want to add "Q1","Q2", etc. to the beginning of question prompts (in a random way). 

What would be the easiest way to achieve this?

Thanks a lot!

**Comment by ChristopheBossens** [comment]
Not sure I completely understand what you are trying to achieve, but I assume that "Q1", "Q2" represents the order in which the questions appear, but the order of the questions themselves is randomized? This is what I have implemented in the example below (for the first presentation). 

The general I would take is to start with an array of questions without the prompt. Then you can shuffle this array to randomize the question order for each presentation. Finally, loop over that new array to update the prompt appropriately. Hope this helps!

```
    // Define a set of questions
    let survey_questions = [
        {
            'prompt' : 'How many cents is each coin worth?',
            'name'   : 'coinWorth',
            'options': ['10','1','0.1','0.01']
        },
        {
            'prompt'  : 'How many cents is each bill worth?',
            'name'    : 'billWorth',
            'options': ['10','1','0.1','0.01']
        },
        {
            'prompt' : 'How many cents is each question worth?',
            'name'   : 'questionWorth',
            'options': ['10','1','0.1','0.01']
        }
    ]

    // Copy and shuffle question order
    let survey_1_questions = jsPsych.randomization.shuffle(survey_questions);

    // Add the appropriate numbers in front of the prompt
    for( let idx = 0; idx < survey_1_questions.length; ++idx){
        survey_1_questions[idx]['prompt'] = 'Q' + (idx + 1).toString() + '. ' + survey_1_questions[idx]['prompt']
    }

    let survey_1 = { 
        type: 'survey-multi-choice', 
        questions: survey_1_questions
	};
```

**Reply by fspinar** [reply]
Thanks @ChristopheBossens ! I tried using your code, but I still have 2 problems:

1st: I want the questions to appear on different pages, but this code puts them in a single page. How can I modify this?

2nd: When I use the above code twice, I got 2 separate Q1, Q2 at the beginning of each question. For ex: 

Code:
`// Add the appropriate numbers in front of the prompt
for( let idx = 0; idx < survey_2_questions.length; ++idx){
  survey_2_questions[idx]['prompt'] = 'Q' + (idx + 1).toString() + '. ' + survey_2_questions[idx]['prompt']
}

let survey_2 = { 
  type: 'survey-multi-choice', 
  questions: survey_2_questions,
};
timeline.push(survey_2);`

Output:

Q10. Q1. How many...
Q2. Q5.  How many...

etc. 
This was supposed to be only Q5. I think this is happening because somehow survey_questions itself is being modified, and when I use the same code again it adds two separate Q2 Q5 to the beginning of each question. I think it has something to do with mutable/immutable objects in JS but I am not very familiar with them.
How can I make it so that there is only one Q at the beginning of each question, no matter how many times I use the survey?

Thanks a lot!!!

**Comment by ChristopheBossens** [comment]
To solve the Q1Q2 problem you need to create a deep copy of the array (forgot about that). Not sure this is the best way, but I found that this is one way to solve it:

```
    // Deep copy and shuffle question order
    let survey_1_questions = JSON.parse(JSON.stringify(survey_questions));
    survey_1_questions = jsPsych.randomization.shuffle(survey_1_questions);
```

To create multiple pages (I assume you mean multiple survey-multi-choice trials?) you first have to calculate the correct number of pages, which you can do once you specify how many questions you want on each page:

```
    // Calculate the number of pages
    let n_questions = survey_1_questions.length;
    let questions_per_page = 1;
    let n_pages = Math.ceil(n_questions / questions_per_page);
```

Finally, you use a loop to add the correct selection of questions to each individual page:

```
    // Create an array with individual pages
    let survey_pages = [];
    for(let page_index = 0; page_index <= n_pages; ++page_index){
        let start_index = page_index * questions_per_page;
        let end_index   = Math.min((page_index + 1) * questions_per_page, n_questions); 
        let selected_questions = survey_1_questions.slice(start_index, end_index);

        let survey_page = { 
            type: 'survey-multi-choice', 
            questions: selected_questions
	    };
        survey_pages.push(survey_page);
    }
```

The individual pages are now in array, so to add them to the timeline you need to unpack the array using the following syntax:
```
    // Add array to the timeline
    jsPsych.init({
        timeline: [...survey_pages]
    })
```
